test/test_cpp_api_parity.py

```diff
@@ -0,0 +1,179 @@
+                for method in torch_nn_test_methods:
+                    functions.append(test_params['module_variant_name'] + '_test_' + method)
+
+        # Just-in-time compile the C++ test code
+        cpp_module = torch.utils.cpp_extension.load_inline(
+            name=test_suite_name,
+            cpp_sources=cpp_source,
+            functions=functions,
+            verbose=True,
+        )
+
+        def test_method(method_name, test_params):
+            expect_error = test_params['expect_error']
+            input_size = test_params['input_size']
+            python_module_class = test_params['python_module_class']
+            python_constructor_args = test_params['python_constructor_args']
+            module_variant_name = test_params['module_variant_name']
+            test_name = module_variant_name + '_test_' + method_name
+            example_input = torch.randn(input_size)
+            with common.freeze_rng_state():
+                torch.manual_seed(2)
+                module = python_module_class(*python_constructor_args)
+                if method_name == 'forward':
+                    python_output = module(example_input)
+                # We use JIT tracing to transfer Python module state to C++
+                traced_script_module = torch.jit.trace(module, example_input)
+                # Ideally we would like to not have to manually delete the file, but NamedTemporaryFile
+                # opens the file, and it cannot be opened multiple times in Windows. To support Windows,
+                # close the file after creation and try to remove it manually
+                f = tempfile.NamedTemporaryFile(delete=False)
+                try:
+                    f.close()
+                    traced_script_module.save(f.name)
+                    torch.manual_seed(2)
+                    if method_name == 'init':
+                        args = (f.name, )
+                    elif method_name == 'forward':
+                        args = (f.name, example_input, python_output)
+                    if expect_error:
+                        with self.assertRaisesRegex(RuntimeError, 'Parity test failed'):
+                            getattr(cpp_module, test_name)(*args)
+                    else:
+                        getattr(cpp_module, test_name)(*args)
+                finally:
+                    os.unlink(f.name)
+
+        for module_name in module_names:
+            for test_params in test_params_map[module_name]:
+                for method in torch_nn_test_methods:
+                    test_method(method, test_params)
+
+    def test_sample_module(self):
+        self._test_torch_nn_modules(
+            test_suite_name='test_sample_module',
+            module_names=['SampleModule'],
+            module_tests=sample_module.module_tests,
+            python_module_class=sample_module.SampleModule,
+            cpp_namespace='')
+
+    def test_torch_nn(self):
```

Comment:
It would be nicer if we generated one test per NN module. This way, we can use the python TestCase infra (and keep getting signals about other tests even if one test fails). Right now it looks like if one nn test fails, then this test fails.